File: tv_alerts/providers/schwab.py
"""Schwab market-data provider.

Schwab's individual developer API uses a standard OAuth2 authorization-code
flow with a short-lived (30 min) access token and a longer-lived (7 day)
refresh token. See SETUP.md for how to register an app and get an App Key
/ App Secret.

This talks directly to the REST API with `requests` rather than pulling in
a third-party Schwab client, so the OAuth/token handling here is fully
visible and auditable. Endpoints and parameter names follow Schwab's
published `marketdata` API (a continuation of the old TD Ameritrade API
shape); double-check against https://developer.schwab.com if Schwab has
changed anything since.
"""

from __future__ import annotations

import json
import os
import stat
import time
import webbrowser
from dataclasses import dataclass
from datetime import date, datetime, timezone
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import requests

from tv_alerts.models import PriceBar

AUTHORIZE_URL = "https://api.schwabapi.com/v1/oauth/authorize"
TOKEN_URL = "https://api.schwabapi.com/v1/oauth/token"
PRICE_HISTORY_URL = "https://api.schwabapi.com/marketdata/v1/pricehistory"

# Refresh a bit before the access token actually expires to avoid racing a
# request against expiry.
_ACCESS_TOKEN_SAFETY_MARGIN_SECONDS = 60


class SchwabAuthError(RuntimeError):
    pass


@dataclass
class _TokenState:
    access_token: str
    refresh_token: str
    access_token_expires_at: float  # unix epoch seconds

    @classmethod
    def from_token_response(cls, payload: dict, obtained_at: float) -> "_TokenState":
        return cls(
            access_token=payload["access_token"],
            refresh_token=payload["refresh_token"],
            access_token_expires_at=obtained_at + payload["expires_in"],
        )

    def to_json(self) -> dict:
        return {
            "access_token": self.access_token,
            "refresh_token": self.refresh_token,
            "access_token_expires_at": self.access_token_expires_at,
        }

    @classmethod
    def from_json(cls, payload: dict) -> "_TokenState":
        return cls(
            access_token=payload["access_token"],
            refresh_token=payload["refresh_token"],
            access_token_expires_at=payload["access_token_expires_at"],
        )


class SchwabAuth:
    """Handles the OAuth2 dance and on-disk token caching for Schwab's API."""

    def __init__(
        self,
        app_key: str,
        app_secret: str,
        token_path: str | Path,
        callback_url: str = "https://127.0.0.1",
    ):
        self.app_key = app_key
        self.app_secret = app_secret
        self.token_path = Path(token_path)
        self.callback_url = callback_url
        self._state: _TokenState | None = self._load_token_state()

    def _load_token_state(self) -> _TokenState | None:
        if not self.token_path.exists():
            return None
        return _TokenState.from_json(json.loads(self.token_path.read_text()))

    def _save_token_state(self) -> None:
        assert self._state is not None
        self.token_path.write_text(json.dumps(self._state.to_json(), indent=2))
        # Token file contains bearer credentials; keep it user-readable only.
        os.chmod(self.token_path, stat.S_IRUSR | stat.S_IWUSR)

    def authorize_interactive(self) -> None:
        """Run the one-time browser login flow and persist the resulting tokens.

        Only needs to be run once per refresh-token lifetime (Schwab refresh
        tokens are valid for 7 days; re-run this whenever get_access_token()
        starts raising SchwabAuthError because the refresh token expired).
        """
        auth_url = (
            f"{AUTHORIZE_URL}?client_id={self.app_key}"
            f"&redirect_uri={self.callback_url}"
        )
        print("Opening Schwab login in your browser. If it doesn't open, visit:")
        print(auth_url)
        webbrowser.open(auth_url)
        redirected_url = input(
            "\nAfter approving access, paste the full URL you were redirected to: "
        ).strip()

        query = parse_qs(urlparse(redirected_url).query)
        if "code" not in query:
            raise SchwabAuthError(
                "Could not find an authorization `code` in the pasted URL."
            )
        code = query["code"][0]

        obtained_at = time.time()
        response = requests.post(
            TOKEN_URL,
            auth=(self.app_key, self.app_secret),
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            data={
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": self.callback_url,
            },
            timeout=30,
        )
        if response.status_code != 200:
            raise SchwabAuthError(
                f"Token exchange failed ({response.status_code}): {response.text}"
            )

        self._state = _TokenState.from_token_response(response.json(), obtained_at)
        self._save_token_state()

    def _refresh(self) -> None:
        if self._state is None:
            raise SchwabAuthError(
                "No cached Schwab tokens found. Run authorize_interactive() first "
                "(e.g. `python -m tv_alerts.cli schwab-login`)."
            )

        obtained_at = time.time()
        response = requests.post(
            TOKEN_URL,
            auth=(self.app_key, self.app_secret),
            headers={"Content-Type": "application/x-www-form-urlencoded"},
            data={
                "grant_type": "refresh_token",
                "refresh_token": self._state.refresh_token,
            },
            timeout=30,
        )
        if response.status_code != 200:
            raise SchwabAuthError(
                f"Token refresh failed ({response.status_code}): {response.text}. "
                "The refresh token may have expired (7 day lifetime) - re-run "
                "authorize_interactive()."
            )

        self._state = _TokenState.from_token_response(response.json(), obtained_at)
        self._save_token_state()

    def get_access_token(self) -> str:
        if self._state is None:
            raise SchwabAuthError(
                "No cached Schwab tokens found. Run authorize_interactive() first "
                "(e.g. `python -m tv_alerts.cli schwab-login`)."
            )
        if time.time() >= self._state.access_token_expires_at - _ACCESS_TOKEN_SAFETY_MARGIN_SECONDS:
            self._refresh()
        return self._state.access_token


class SchwabProvider:
    """`PriceDataProvider` backed by Schwab's `/marketdata/v1/pricehistory`."""

    def __init__(self, auth: SchwabAuth, session: requests.Session | None = None):
        self.auth = auth
        self.session = session or requests.Session()

    def get_daily_bars(self, symbol: str, start: date, end: date) -> list[PriceBar]:
        start_ms = int(
            datetime(start.year, start.month, start.day, tzinfo=timezone.utc).timestamp()
            * 1000
        )
        end_ms = int(
            datetime(end.year, end.month, end.day, tzinfo=timezone.utc).timestamp()
            * 1000
        )

        response = self.session.get(
            PRICE_HISTORY_URL,
            headers={"Authorization": f"Bearer {self.auth.get_access_token()}"},
            params={
                "symbol": symbol,
                "periodType": "year",
                "frequencyType": "daily",
                "frequency": 1,
                "startDate": start_ms,
                "endDate": end_ms,
                "needExtendedHoursData": "false",
            },
            timeout=30,
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"Schwab price history request for {symbol} failed "
                f"({response.status_code}): {response.text}"
            )

        payload = response.json()
        bars = []
        for candle in payload.get("candles", []):
            bars.append(
                PriceBar(
                    date=datetime.fromtimestamp(candle["datetime"] / 1000, tz=timezone.utc),
                    open=candle["open"],
                    high=candle["high"],
                    low=candle["low"],
                    close=candle["close"],
                    volume=int(candle["volume"]),
                )
            )
        return bars

File: src/providers/schwab.ts
/**
 * Schwab market-data provider.
 *
 * Schwab's individual developer API uses a standard OAuth2 authorization-code
 * flow with a short-lived (30 min) access token and a longer-lived (7 day)
 * refresh token. See SETUP.md for how to register an app and get an App Key
 * / App Secret.
 *
 * This talks directly to the REST API with `fetch` rather than pulling in a
 * third-party Schwab client, so the OAuth/token handling here is fully
 * visible and auditable. Endpoints and parameter names follow Schwab's
 * published `marketdata` API (a continuation of the old TD Ameritrade API
 * shape); double-check against https://developer.schwab.com if Schwab has
 * changed anything since.
 */

import { chmodSync, existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { createInterface } from "node:readline/promises";
import open from "open";
import type { PriceBar } from "../models.js";
import { RateLimiter } from "./rateLimiter.js";
import type { PriceDataProvider } from "./types.js";

const AUTHORIZE_URL = "https://api.schwabapi.com/v1/oauth/authorize";
const TOKEN_URL = "https://api.schwabapi.com/v1/oauth/token";
const PRICE_HISTORY_URL = "https://api.schwabapi.com/marketdata/v1/pricehistory";
const QUOTES_URL = "https://api.schwabapi.com/marketdata/v1/quotes";

// Refresh a bit before the access token actually expires to avoid racing a
// request against expiry.
const ACCESS_TOKEN_SAFETY_MARGIN_SECONDS = 60;

export class SchwabAuthError extends Error {}

interface TokenState {
  accessToken: string;
  refreshToken: string;
  accessTokenExpiresAt: number; // unix epoch seconds
}

interface TokenResponse {
  access_token: string;
  refresh_token: string;
  expires_in: number;
}

function tokenStateFromResponse(payload: TokenResponse, obtainedAt: number): TokenState {
  return {
    accessToken: payload.access_token,
    refreshToken: payload.refresh_token,
    accessTokenExpiresAt: obtainedAt + payload.expires_in,
  };
}

export class SchwabAuth {
  private appKey: string;
  private appSecret: string;
  private tokenPath: string;
  private callbackUrl: string;
  private state: TokenState | null;

  constructor(appKey: string, appSecret: string, tokenPath: string, callbackUrl = "https://127.0.0.1") {
    this.appKey = appKey;
    this.appSecret = appSecret;
    this.tokenPath = tokenPath;
    this.callbackUrl = callbackUrl;
    this.state = this.loadTokenState();
  }

  private loadTokenState(): TokenState | null {
    if (!existsSync(this.tokenPath)) {
      return null;
    }
    return JSON.parse(readFileSync(this.tokenPath, "utf-8")) as TokenState;
  }

  private saveTokenState(): void {
    if (!this.state) {
      return;
    }
    mkdirSync(dirname(this.tokenPath), { recursive: true });
    writeFileSync(this.tokenPath, JSON.stringify(this.state, null, 2));
    // Token file contains bearer credentials; keep it user-readable only.
    chmodSync(this.tokenPath, 0o600);
  }

  private basicAuthHeader(): string {
    return "Basic " + Buffer.from(`${this.appKey}:${this.appSecret}`).toString("base64");
  }

  /**
   * Run the one-time browser login flow and persist the resulting tokens.
   *
   * Only needs to be run once per refresh-token lifetime (Schwab refresh
   * tokens are valid for 7 days; re-run this whenever getAccessToken() starts
   * throwing SchwabAuthError because the refresh token expired).
   */
  async authorizeInteractive(): Promise<void> {
    const authUrl = `${AUTHORIZE_URL}?client_id=${this.appKey}&redirect_uri=${this.callbackUrl}`;
    console.log("Opening Schwab login in your browser. If it doesn't open, visit:");
    console.log(authUrl);
    await open(authUrl);

    const rl = createInterface({ input: process.stdin, output: process.stdout });
    const redirectedUrl = (
      await rl.question("\nAfter approving access, paste the full URL you were redirected to: ")
    ).trim();
    rl.close();

    const code = new URL(redirectedUrl).searchParams.get("code");
    if (!code) {
      throw new SchwabAuthError("Could not find an authorization `code` in the pasted URL.");
    }

    const obtainedAt = Date.now() / 1000;
    const response = await fetch(TOKEN_URL, {
      method: "POST",
      headers: {
        Authorization: this.basicAuthHeader(),
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        grant_type: "authorization_code",
        code,
        redirect_uri: this.callbackUrl,
      }),
    });
    if (!response.ok) {
      throw new SchwabAuthError(`Token exchange failed (${response.status}): ${await response.text()}`);
    }

    this.state = tokenStateFromResponse((await response.json()) as TokenResponse, obtainedAt);
    this.saveTokenState();
  }

  private async refresh(): Promise<void> {
    if (!this.state) {
      throw new SchwabAuthError(
        "No cached Schwab tokens found. Run authorizeInteractive() first " +
          "(e.g. `tv-alerts schwab-login`)."
      );
    }

    const obtainedAt = Date.now() / 1000;
    const response = await fetch(TOKEN_URL, {
      method: "POST",
      headers: {
        Authorization: this.basicAuthHeader(),
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        grant_type: "refresh_token",
        refresh_token: this.state.refreshToken,
      }),
    });
    if (!response.ok) {
      throw new SchwabAuthError(
        `Token refresh failed (${response.status}): ${await response.text()}. ` +
          "The refresh token may have expired (7 day lifetime) - re-run authorizeInteractive()."
      );
    }

    this.state = tokenStateFromResponse((await response.json()) as TokenResponse, obtainedAt);
    this.saveTokenState();
  }

  async getAccessToken(): Promise<string> {
    if (!this.state) {
      throw new SchwabAuthError(
        "No cached Schwab tokens found. Run authorizeInteractive() first " + "(e.g. `tv-alerts schwab-login`)."
      );
    }
    if (Date.now() / 1000 >= this.state.accessTokenExpiresAt - ACCESS_TOKEN_SAFETY_MARGIN_SECONDS) {
      await this.refresh();
    }
    return this.state!.accessToken;
  }
}

interface Candle {
  datetime: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

interface PriceHistoryResponse {
  candles: Candle[];
}

interface QuoteResponse {
  [symbol: string]: { quote?: { lastPrice: number } };
}

// Schwab's Market Data API is limited to 120 calls/minute per app.
export const DEFAULT_MAX_REQUESTS_PER_MINUTE = 120;

export class SchwabProvider implements PriceDataProvider {
  private rateLimiter: RateLimiter;

  constructor(
    private auth: SchwabAuth,
    maxRequestsPerMinute: number = DEFAULT_MAX_REQUESTS_PER_MINUTE
  ) {
    this.rateLimiter = new RateLimiter(maxRequestsPerMinute, 60_000);
  }

  async getDailyBars(symbol: string, start: Date, end: Date): Promise<PriceBar[]> {
    const params = new URLSearchParams({
      symbol,
      periodType: "year",
      frequencyType: "daily",
      frequency: "1",
      startDate: String(start.getTime()),
      endDate: String(end.getTime()),
      needExtendedHoursData: "false",
    });

    await this.rateLimiter.acquire();
    const accessToken = await this.auth.getAccessToken();
    const response = await fetch(`${PRICE_HISTORY_URL}?${params}`, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });
    if (!response.ok) {
      throw new Error(
        `Schwab price history request for ${symbol} failed (${response.status}): ${await response.text()}`
      );
    }

    const payload = (await response.json()) as PriceHistoryResponse;
    return (payload.candles ?? []).map((candle) => ({
      date: new Date(candle.datetime),
      open: candle.open,
      high: candle.high,
      low: candle.low,
      close: candle.close,
      volume: candle.volume,
    }));
  }

  /** Current last-traded price per symbol, for periodic alert checks. */
  async getQuotes(symbols: string[]): Promise<Map<string, number>> {
    if (symbols.length === 0) {
      return new Map();
    }

    await this.rateLimiter.acquire();
    const accessToken = await this.auth.getAccessToken();
    const params = new URLSearchParams({ symbols: symbols.join(",") });
    const response = await fetch(`${QUOTES_URL}?${params}`, {
      headers: { Authorization: `Bearer ${accessToken}` },
    });
    if (!response.ok) {
      throw new Error(`Schwab quotes request failed (${response.status}): ${await response.text()}`);
    }

    const payload = (await response.json()) as QuoteResponse;
    const result = new Map<string, number>();
    for (const [symbol, data] of Object.entries(payload)) {
      if (data.quote?.lastPrice !== undefined) {
        result.set(symbol, data.quote.lastPrice);
      }
    }
    return result;
  }
}
